libs/experimental/langchain_experimental/data_anonymizer/presidio.py

```diff
@@ -89,3 +94,133 @@ def add_recognizer(self, recognizer: EntityRecognizer) -> None:
+                "`pip install presidio-anonymizer`."
+            ) from e
+
+        self.analyzed_fields = (
+            analyzed_fields
+            if analyzed_fields is not None
+            else list(get_pseudoanonymizer_mapping().keys())
+        )
+        self.language = language
+        self.operators = (
+            operators
+            if operators is not None
+            else {
+                field: OperatorConfig(
+                    operator_name="custom", params={"lambda": faker_function}
+                )
+                for field, faker_function in get_pseudoanonymizer_mapping().items()
+            }
+        )
+        self._analyzer = AnalyzerEngine()
+        self._anonymizer = AnonymizerEngine()
+        self.deanonymizer_mapping: Dict[str, Dict[str, str]] = dict()
+
+    def _create_update_deanonymizer_mapping(
+        self,
+        original_text: str,
+        analyzer_results: List[RecognizerResult],
+        anonymizer_results: EngineResult,
+    ) -> None:
+        """Create a mapping from anonymized text to original text"""
+        analyzer_results = sorted(analyzer_results, key=lambda d: d.start)
+        anonymizer_results.items = sorted(
+            anonymizer_results.items, key=lambda d: d.start
+        )
+
+        new_deanonymizer_mapping: Dict[str, Dict[str, str]] = {}
+
+        for analyzed_entity, anonymized_entity in zip(
+            analyzer_results, anonymizer_results.items
+        ):
+            original_value = original_text[analyzed_entity.start : analyzed_entity.end]
+
+            if anonymized_entity.entity_type not in new_deanonymizer_mapping:
+                new_deanonymizer_mapping[anonymized_entity.entity_type] = {}
+
+            new_deanonymizer_mapping[anonymized_entity.entity_type][
+                anonymized_entity.text
+            ] = original_value
+
+        for entity_type, values in new_deanonymizer_mapping.items():
+            if entity_type not in self.deanonymizer_mapping:
+                self.deanonymizer_mapping[entity_type] = {}
+
+            self.deanonymizer_mapping[entity_type].update(values)
+
+    def _anonymize(self, text: str) -> str:
+        analyzer_results = self._analyzer.analyze(
+            text,
+            entities=self.analyzed_fields,
+            language=self.language,
+        )
+
+        filtered_analyzer_results = (
+            self._anonymizer._remove_conflicts_and_get_text_manipulation_data(
+                analyzer_results
+            )
+        )
+
+        anonymizer_results = self._anonymizer.anonymize(
+            text,
+            analyzer_results=analyzer_results,
+            operators=self.operators,
+        )
+
+        self._create_update_deanonymizer_mapping(
+            text, filtered_analyzer_results, anonymizer_results
+        )
+
+        return anonymizer_results.text
+
+    def _deanonymize(self, text_to_deanonymize: str) -> str:
```

Comment:
probably we should warn the user if he wants to deanonymize data before any anonimization was done